File: src/skills.ts
import { readdir } from "node:fs/promises";
import { join } from "node:path";
import { homedir } from "node:os";

export type Skill = {
  name: string;
  description: string;
  path: string;
  content: string;
};

// ── P1 skills cache ─────────────────────────────────────────────────────────
let _cached: Skill[] | null = null;
let _cachedAt = 0;
const CACHE_TTL_MS = 60_000;
// ─────────────────────────────────────────────────────────────────────────────

function parseFrontmatter(raw: string): { name?: string; description?: string; body: string } {
  if (!raw.startsWith("---")) return { body: raw };
  const end = raw.indexOf("\n---", 3);
  if (end === -1) return { body: raw };
  const fm = raw.slice(3, end).trim();
  const body = raw.slice(end + 4).trimStart(); // skip \n---
  let name: string | undefined;
  let description: string | undefined;
  for (const line of fm.split("\n")) {
    const m = line.match(/^\s*(name|description)\s*:\s*(.*)\s*$/);
    if (!m) continue;
    const key = m[1]!.trim();
    let val = m[2]!.trim();
    // strip quotes
    if ((val.startsWith('"') && val.endsWith('"')) || (val.startsWith("'") && val.endsWith("'"))) {
      val = val.slice(1, -1);
    }
    if (key === "name") name = val;
    if (key === "description") description = val;
  }
  return { name, description, body };
}

async function scanDir(base: string): Promise<Skill[]> {
  const skills: Skill[] = [];
  try {
    const entries = await readdir(base, { withFileTypes: true });
    for (const ent of entries) {
      if (!ent.isDirectory()) continue;
      const skillPath = join(base, ent.name, "SKILL.md");
      const file = Bun.file(skillPath);
      if (!(await file.exists())) continue;
      const raw = await file.text();
      const { name, description, body } = parseFrontmatter(raw);
      const skillName = name || ent.name;
      const desc = description || body.split("\n").find((l) => l.trim())?.slice(0, 140) || "";
      skills.push({
        name: skillName,
        description: desc,
        path: skillPath,
        content: raw,
      });
    }
  } catch {
    // dir doesn't exist or unreadable — ignore
  }
  return skills;
}

export async function discoverSkills(): Promise<Skill[]> {
  // Return cached result if still fresh
  if (_cached && Date.now() - _cachedAt < CACHE_TTL_MS) return _cached;

  const home = homedir();
  const bases = [
    join(process.cwd(), "skills"),
    join(process.cwd(), ".lean", "skills"),
    join(home, ".agents", "skills"),
  ];

  const all: Map<string, Skill> = new Map();
  for (const base of bases) {
    const found = await scanDir(base);
    for (const s of found) {
      // local project dirs win over general ~/.agents/skills
      const isLocal = base.startsWith(process.cwd());
      if (!all.has(s.name) || isLocal) {
        all.set(s.name, s);
      }
    }
  }
  const result = [...all.values()].sort((a, b) => a.name.localeCompare(b.name));

  // Store in cache
  _cached = result;
  _cachedAt = Date.now();
  return result;
}

export async function loadSkill(name: string): Promise<string> {
  const skills = await discoverSkills();
  const found = skills.find((s) => s.name === name);
  if (!found) {
    const avail = skills.map((s) => s.name).join(", ") || "(none)";
    throw new Error(`Skill not found: ${name}. Available: ${avail}`);
  }
  return found.content;
}

export async function getSkillCatalog(): Promise<string> {
  const skills = await discoverSkills();
  // Filter pi-internal meta-skills that should not auto-trigger for every prompt
  const filtered = skills.filter((s) => s.name !== "using-superpowers");
  if (filtered.length === 0) return "No skills installed. Use `npx skills add <skill>` to install to ~/.agents/skills.";
  return filtered
    .map((s) => `- ${s.name}: ${s.description} (path: ${s.path})`)
    .join("\n");
}

File: src/skills.rs
use std::collections::HashMap;
use std::path::{Path, PathBuf};
use std::sync::OnceLock;
use std::time::{Duration, Instant};

use anyhow::{Context, Result};
use dirs::home_dir;
use walkdir::WalkDir;

const CACHE_TTL: Duration = Duration::from_secs(60);

#[derive(Debug, Clone)]
pub struct Skill {
    pub name: String,
    pub description: String,
    pub path: PathBuf,
    pub content: String,
}

// Cache state
static CACHE: OnceLock<std::sync::Mutex<Option<(Vec<Skill>, Instant)>>> = OnceLock::new();

fn cache_lock() -> &'static std::sync::Mutex<Option<(Vec<Skill>, Instant)>> {
    CACHE.get_or_init(|| std::sync::Mutex::new(None))
}

fn parse_frontmatter(raw: &str) -> (Option<String>, Option<String>, String) {
    if !raw.starts_with("---") {
        return (None, None, raw.to_string());
    }
    let end = raw[3..].find("\n---").map(|i| i + 3);
    let end = match end {
        Some(i) => i,
        None => return (None, None, raw.to_string()),
    };
    let fm = &raw[3..end].trim();
    let body = raw[end + 4..].trim_start();

    let mut name = None;
    let mut description = None;
    for line in fm.lines() {
        if let Some((key, val)) = line.split_once(':') {
            let key = key.trim();
            let val = val.trim();
            let val = val
                .strip_prefix('"')
                .and_then(|v| v.strip_suffix('"'))
                .or_else(|| val.strip_prefix('\'').and_then(|v| v.strip_suffix('\'')))
                .unwrap_or(val)
                .trim();
            match key {
                "name" => name = Some(val.to_string()),
                "description" => description = Some(val.to_string()),
                _ => {}
            }
        }
    }
    (name, description, body.to_string())
}

async fn scan_dir(base: &Path) -> Vec<Skill> {
    let mut skills = Vec::new();
    let read_dir = match tokio::fs::read_dir(base).await {
        Ok(rd) => rd,
        Err(_) => return skills,
    };
    let mut entries = read_dir;
    while let Ok(Some(entry)) = entries.next_entry().await {
        let file_type = match entry.file_type().await {
            Ok(ft) => ft,
            Err(_) => {
                if let Ok(meta) = std::fs::metadata(entry.path()) {
                    meta.file_type()
                } else {
                    continue;
                }
            }
        };
        if !file_type.is_dir() {
            continue;
        }
        let skill_path = entry.path().join("SKILL.md");
        if !skill_path.exists() {
            continue;
        }
        let raw = match tokio::fs::read_to_string(&skill_path).await {
            Ok(c) => c,
            Err(_) => continue,
        };
        let (name_opt, description_opt, body) = parse_frontmatter(&raw);
        let skill_name = name_opt.unwrap_or_else(|| entry.file_name().to_string_lossy().to_string());
        let desc = description_opt.unwrap_or_else(|| {
            body.lines()
                .find(|l| !l.trim().is_empty())
                .unwrap_or("")
                .chars()
                .take(140)
                .collect()
        });
        skills.push(Skill {
            name: skill_name,
            description: desc,
            path: skill_path,
            content: raw,
        });
    }
    skills
}

pub async fn discover_skills() -> Vec<Skill> {
    {
        let mut lock = cache_lock().lock().unwrap();
        if let Some((cached, time)) = lock.as_ref() {
            if time.elapsed() < CACHE_TTL {
                return cached.clone();
            }
        }
    }

    let cwd = std::env::current_dir().unwrap_or_else(|_| PathBuf::from("."));
    let home = home_dir().unwrap_or_else(|| PathBuf::from("~"));

    let bases = vec![
        cwd.join("skills"),
        cwd.join(".lean").join("skills"),
        home.join(".agents").join("skills"),
    ];

    let mut all: HashMap<String, Skill> = HashMap::new();
    for base in &bases {
        let found = scan_dir(base).await;
        let is_local = base.starts_with(&cwd);
        for skill in found {
            if !all.contains_key(&skill.name) || is_local {
                all.insert(skill.name.clone(), skill);
            }
        }
    }
    let mut result: Vec<Skill> = all.into_values().collect();
    result.sort_by(|a, b| a.name.cmp(&b.name));

    {
        let mut lock = cache_lock().lock().unwrap();
        *lock = Some((result.clone(), Instant::now()));
    }

    result
}

pub async fn get_skill_catalog() -> String {
    let skills = discover_skills().await;
    // Filter pi-internal meta-skills that should not auto-trigger for every prompt
    let filtered: Vec<&Skill> = skills.iter().filter(|s| s.name != "using-superpowers").collect();
    if filtered.is_empty() {
        return "No skills installed. Use `pi skill add <skill>` to install to ~/.agents/skills.".to_string();
    }
    filtered
        .iter()
        .map(|s| format!("- {}: {} (path: {})", s.name, s.description, s.path.display()))
        .collect::<Vec<_>>()
        .join("\n")
}

pub async fn load_skill(name: &str) -> Result<String> {
    let skills = discover_skills().await;
    let found = skills.iter().find(|s| s.name == name);
    match found {
        Some(skill) => Ok(skill.content.clone()),
        None => {
            let avail: Vec<&str> = skills.iter().map(|s| s.name.as_str()).collect();
            anyhow::bail!(
                "Skill not found: {}. Available: {}",
                name,
                avail.join(", ")
            );
        }
    }
}